client/packages/core/src/queryTypes.ts

```diff
@@ -42,20 +45,127 @@ type WhereArgs = {
   $ilike?: string;
 };
 
-type WhereClauseValue = string | number | boolean | NonEmpty<WhereArgs>;
+type BaseWhereClauses<V> = {
+  in?: V[];
+  $in?: V[];
+  $not?: V;
+  $gt?: V;
+  $lt?: V;
+  $gte?: V;
+  $lte?: V;
+};
 
-type BaseWhereClause = {
-  [key: string]: WhereClauseValue;
+type WhereArgs<V> = V extends BSUnknown
+  ? AnyWhereArgs
+  : BaseWhereClauses<V> &
+      (V extends string
+        ? {
+            $like?: string;
+            $ilike?: string;
+          }
+        : {}) &
+      (undefined extends V
+        ? {
+            $isNull?: boolean;
+          }
+        : {});
+
+// Make type display better
+type WhereClauseValue<V> =
+  | (V extends BSUnknown
+      ? string | number | boolean
+      : V extends string | undefined
+        ? string
+        : V extends number | undefined
+          ? number
+          : V extends boolean | undefined
+            ? boolean
+            : any)
+  | Expand<WhereArgs<V>>;
+
+type BaseWhereClause<
+  T extends {
+    [key: string]: unknown;
+  },
+> = {
+  [key in keyof T]?: WhereClauseValue<T[key]>;
 };
 
-type WhereClauseWithCombination = {
-  or?: WhereClause[] | WhereClauseValue;
-  and?: WhereClause[] | WhereClauseValue;
+// Helper type to infer the value type from a nested path
+type InferNestedValueType<
+  S extends IContainEntitiesAndLinks<any, any>,
+  K extends keyof S['entities'],
+  Path extends string,
+  Depth extends number = 3,
+> = Depth extends 0
+  ? BSUnknown
```

Comment:
I assume Depth is counting down? I was a bit surprised, as I would have expected the number to count up